addons/l10n_in/tests/test_invoice_hsn_summary.py

```diff
@@ -0,0 +1,624 @@
+        super().setUpClass(chart_template_ref='in')
+
+        cls.test_hsn_code_1 = '1234'
+        cls.test_hsn_code_2 = '4321'
+
+        cls.uom_unit = cls.env.ref('uom.product_uom_unit')
+        cls.uom_dozen = cls.env.ref('uom.product_uom_dozen')
+
+        cls.product_a.l10n_in_hsn_code = cls.test_hsn_code_1
+        cls.product_b.l10n_in_hsn_code = cls.test_hsn_code_2
+        cls.product_c = cls.env['product.product'].create({
+            'name': 'product_c',
+            'l10n_in_hsn_code': cls.test_hsn_code_1,
+            'uom_id': cls.env.ref('uom.product_uom_unit').id,
+            'lst_price': 1000.0,
+            'property_account_income_id': cls.company_data['default_account_revenue'].id,
+        })
+
+        cls.gst_5 = cls.env['account.chart.template'].ref('sgst_sale_5')
+        cls.gst_18 = cls.env['account.chart.template'].ref('sgst_sale_18')
+
+        cls.igst_0 = cls.env['account.chart.template'].ref('igst_sale_0')
+        cls.igst_5 = cls.env['account.chart.template'].ref('igst_sale_5')
+        cls.igst_18 = cls.env['account.chart.template'].ref('igst_sale_18')
+        cls.cess_5_plus_1591 = cls.env['account.chart.template'].ref('cess_5_plus_1591_sale')
+        cls.exempt_0 = cls.env['account.chart.template'].ref('exempt_sale')
+
+        cls.display_uom = cls.env.user.user_has_groups('uom.group_uom')
+
+    def assert_hsn_summary(self, invoice, expected_values):
+        hsn_summary = invoice._l10n_in_get_hsn_summary_table()
+        self.assertEqual(
+            {k: len(v) if k == 'items' else v for k, v in hsn_summary.items()},
+            {k: len(v) if k == 'items' else v for k, v in expected_values.items()},
+        )
+        self.assertEqual(len(hsn_summary['items']), len(expected_values['items']))
+        for item, expected_item in zip(hsn_summary['items'], expected_values['items']):
+            self.assertDictEqual(item, expected_item)
+
+    @freeze_time('2019-01-01')
+    def create_invoice(self, invoice_line_ids, **kwargs):
+        return self.env['account.move'].create({
+            'move_type': 'out_invoice',
+            'partner_id': self.partner_a.id,
+            **kwargs,
+            'invoice_line_ids': invoice_line_ids,
+        })
+
+    def test_l10n_in_hsn_summary_1(self):
+        """ Test GST/IGST taxes. """
+        invoice = self.create_invoice(invoice_line_ids=[
+            Command.create({
+                'product_id': self.product_a.id,
+                'quantity': 2.0,
+                'price_unit': 100,
+                'tax_ids': [Command.set(self.gst_5.ids)], #Tax: 10
+            }),
+            Command.create({
+                'product_id': self.product_c.id,
+                'quantity': 1.0,
+                'price_unit': 600,
+                'tax_ids': [Command.set(self.gst_5.ids)], #Tax: 30
+            }),
+            Command.create({
+                'product_id': self.product_a.id,
+                # 'l10n_in_hsn_code': self.test_hsn_code_1,
```

Comment:
Made a lazy fix as of now, it will require less rework in fw-port may be I can remove this in my l10n_in test case PR which already exist in 17.0